File: backend/routes/alerts.js
const express = require('express');
const router = express.Router();
const Alert = require('../models/Alert');
const { InMemoryStore } = require('../seed/seedData');
const { getIsMock } = require('../config/db');
const wsManager = require('../services/websocket');

// 1. GET /api/alerts
router.get('/', async (req, res) => {
  try {
    const isMock = getIsMock();
    let alerts = [];

    if (!isMock) {
      alerts = await Alert.find({}).sort({ dispatched_at: -1 }).limit(100).catch(() => []);
    }

    if (!alerts || alerts.length === 0) {
      alerts = InMemoryStore.alerts;
    }

    return res.json(alerts);
  } catch (err) {
    return res.json(InMemoryStore.alerts);
  }
});

// 2. POST /api/alerts/:id/cancel
router.post('/:id/cancel', async (req, res) => {
  try {
    const { id } = req.params;
    const isMock = getIsMock();

    if (!isMock) {
      await Alert.updateOne(
        { _id: id },
        { $set: { resolved: true, resolved_at: new Date() } }
      ).catch(() => null);
    }

    const memAlert = InMemoryStore.alerts.find(a => String(a._id) === id);
    if (memAlert) {
      memAlert.resolved = true;
      memAlert.resolved_at = new Date();
    }

    wsManager.broadcast({
      type: 'alert_cancelled',
      alert_id: id
    });

    return res.json({ status: 'cancelled', alert_id: id });
  } catch (err) {
    return res.status(500).json({ status: 'error', message: err.message });
  }
});

// 3. POST /api/alerts/:id/mark-false
router.post('/:id/mark-false', async (req, res) => {
  try {
    const { id } = req.params;
    const isMock = getIsMock();

    if (!isMock) {
      await Alert.updateOne(
        { _id: id },
        { $set: { false_alarm: true, resolved: true, resolved_at: new Date() } }
      ).catch(() => null);
    }

    const memAlert = InMemoryStore.alerts.find(a => String(a._id) === id);
    if (memAlert) {
      memAlert.false_alarm = true;
      memAlert.resolved = true;
      memAlert.resolved_at = new Date();
    }

    return res.json({ status: 'marked_false', alert_id: id });
  } catch (err) {
    return res.status(500).json({ status: 'error', message: err.message });
  }
});

module.exports = router;

File: backend/routes/alerts.py
from fastapi import APIRouter, Depends, HTTPException, status
from typing import List
from database import get_db
from models.alert import Alert
from routes.auth import get_current_user
from datetime import datetime
from bson import ObjectId

router = APIRouter()

@router.get("/", response_model=List[Alert])
async def get_alerts(current_user = Depends(get_current_user)):
    db = get_db()
    alerts = await db.alerts.find().sort("dispatched_at", -1).to_list(1000)
    # Convert ObjectIds to strings if necessary
    for alert in alerts:
        if "_id" in alert and isinstance(alert["_id"], ObjectId):
            alert["_id"] = str(alert["_id"])
    return alerts

@router.post("/{alert_id}/cancel")
async def cancel_alert(alert_id: str, current_user = Depends(get_current_user)):
    db = get_db()
    try:
        obj_id = ObjectId(alert_id)
    except:
        obj_id = alert_id
        
    result = await db.alerts.update_one(
        {"_id": obj_id},
        {"$set": {"resolved": True, "resolved_at": datetime.utcnow()}}
    )
    if result.modified_count == 0:
        raise HTTPException(status_code=404, detail="Alert not found")
    
    from services.websocket_manager import manager
    await manager.broadcast({
        "type": "alert_cancelled",
        "alert_id": alert_id
    })
    return {"status": "cancelled"}

@router.post("/{alert_id}/mark-false")
async def mark_false_alarm(alert_id: str, current_user = Depends(get_current_user)):
    db = get_db()
    try:
        obj_id = ObjectId(alert_id)
    except:
        obj_id = alert_id
        
    result = await db.alerts.update_one(
        {"_id": obj_id},
        {"$set": {"false_alarm": True, "resolved": True, "resolved_at": datetime.utcnow()}}
    )
    if result.modified_count == 0:
        raise HTTPException(status_code=404, detail="Alert not found")
        
    return {"status": "marked_false"}
